workers/unity/Assets/Playground/Scripts/EntityInitialization/EntityGameObjectCreator.cs

```diff
@@ -1,46 +1,67 @@
 {
-    public class EntityGameObjectCreator
+    public class EntityGameObjectCreator : IEntityGameObjectCreator
     {
-        private readonly World world;
-        private readonly Dictionary<string, GameObject> cachedPrefabs;
+        private readonly Dictionary<string, Dictionary<string, GameObject>> cachedPrefabsPerWorkerType
+            = new Dictionary<string, Dictionary<string, GameObject>>();
 
-        public EntityGameObjectCreator(World world, Dictionary<string, GameObject> initialCachedPrefabs = null)
+        public GameObject CreateGameObjectForEntity(SpatialOSEntity entity, WorkerSystem worker)
         {
-            this.world = world;
-            cachedPrefabs = initialCachedPrefabs ?? new Dictionary<string, GameObject>();
-        }
+            if (!cachedPrefabsPerWorkerType.TryGetValue(worker.WorkerType, out var cachedPrefabs))
+            {
+                cachedPrefabsPerWorkerType[worker.WorkerType] = cachedPrefabs = new Dictionary<string, GameObject>();
+            }
 
-        public GameObject CreateEntityGameObject(Entity entity, string prefabPath, Vector3 position,
-            Quaternion rotation, EntityId spatialEntityId)
-        {
-            if (!cachedPrefabs.TryGetValue(prefabPath, out var prefab))
+            if (!entity.HasComponent<Metadata.Component>())
+            {
+                return null;
+            }
+
+            var prefabName = entity.GetComponent<Metadata.Component>().EntityType;
+            if (!entity.HasComponent<Transform.Component>())
             {
-                prefab = Resources.Load<GameObject>(prefabPath);
+                cachedPrefabs[prefabName] = null;
+                return null;
+            }
+
+            var transform = entity.GetComponent<Transform.Component>();
+            var position = new Vector3(transform.Location.X, transform.Location.Y, transform.Location.Z) +
+                worker.Origin;
+            var rotation = new UnityEngine.Quaternion(transform.Rotation.X, transform.Rotation.Y,
+                transform.Rotation.Z, transform.Rotation.W);
+            var workerSpecificPath = Path.Combine("Prefabs", worker.WorkerType, prefabName);
+            var commonPath = Path.Combine("Prefabs", "Common", prefabName);
+
+            if (!cachedPrefabs.TryGetValue(workerSpecificPath, out var prefab)
+                && !cachedPrefabs.TryGetValue(commonPath, out prefab))
+            {
+                prefab = Resources.Load<GameObject>(workerSpecificPath);
                 if (prefab == null)
                 {
-                    throw new PrefabNotFoundException($"Prefab for prefabPath {prefabPath} not found.");
+                    prefab = Resources.Load<GameObject>(commonPath);
                 }
 
-                cachedPrefabs[prefabPath] = prefab;
+                // Cache even if null
+                cachedPrefabs[prefabName] = prefab;
+            }
+
+            if (prefab == null)
+            {
+                return null;
```

Comment:
we should log it, if it can't find a prefab

Reply:
It's not logged at the moment because it's supposed to mean the user does not want a GameObject for that Spatial entity, but it's definitely an annoying error case. Probably best to log once when failing to load, we can tell that first attempt apart using the cache anyway, I'll change that

Reply:
Will go with logging a warning in these cases, sounds OK?

Reply:
Actually going with standard logging instead because it's expected to happen as normal use